src/libsyntax/codemap.rs

```diff
@@ -807,6 +807,83 @@ impl CodeMap {
                         hi.col.to_usize() + 1)).to_string()
     }
 
+    // Returns true if two spans have the same callee
+    // (Assumes the same ExpnFormat implies same callee)
+    fn match_callees(&self, sp_a: &Span, sp_b: &Span) -> bool {
+        let fmt_a = self
+            .with_expn_info(sp_a.expn_id,
+                            |ei| ei.map(|ei| ei.callee.format.clone()));
+
+        let fmt_b = self
+            .with_expn_info(sp_b.expn_id,
+                            |ei| ei.map(|ei| ei.callee.format.clone()));
+        fmt_a == fmt_b
+    }
+
+    /// Returns a formatted string showing the expansion chain of a span
+    ///
+    /// Spans are printed in the following format:
+    ///
+    /// filename:start_line:col: end_line:col
+    /// snippet
+    ///   CALLEE:
+    ///   Callee span, or 'None'
+    ///   CALLSITE:
+    ///   Callsite span, or 'None'
+    ///
+    /// Callees and callsites are printed recursively, expanding into their
+    /// own callee/callsite spans. Each layer of recursion has an increased indent
+    pub fn span_to_expanded_string(&self, sp:Span, dent: &str) -> String {
+        let mut indent = dent.to_owned();
+        let mut output = "".to_owned();
+        let src_str = self.span_to_string(sp);
+        let src_snip = self.span_to_snippet(sp)
+            .unwrap_or("Snippet unavailable (span malformed or source missing)".to_owned());
+        output.push_str(&format!("{}{}\n{}{}\n", indent, src_str, indent, src_snip));
+
+        if sp.expn_id != NO_EXPANSION && sp.expn_id != COMMAND_LINE_EXPN {
+            let mut callee = self.with_expn_info(sp.expn_id,
+                                             |ei| ei.map_or(None, |ei| ei.callee.span.clone()));
```

Comment:
Because callee spans are an Option Span. So to save having an Option<Option<Span>>, I use map_or so I can have an Option<Span> and treat it in a way consistent with callsite.